### PR TITLE
Niraj:Narcissistic number computation, test coverage implementation

### PR DESCRIPTION
**Pull request by nirajunify**
This commit/PR consists of:
- Narcissistic numbers computation and test
- Added Code coverage configuration in jest.config.js
- Added coverage directory in .gitignored
- Added source map support in .tsconfig.json
- Added Jest Config article link in references
- Removed dummy foo.ts and its test file

### CODE REVIEW

File: src/niraj/narcissisticNumbers.spec.ts [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+});
+
+describe('Given a negative range', () => {
+  test('it should return an empty array', () => {
+    expect(narcissisticNumbers(-20, -10).length).toBe(0);
+  });
+});
+
+describe('When no range given', () => {
+  test('it should return an empty array', () => {
```
Comment (krajasekhar-unify) [inline src/niraj/narcissisticNumbers.spec.ts]: Isn't the purpose of `describe` is to group the tests and have some common scope and it's own `beforeEach` and `afterEach` than using it for each tests. Having `describe` for each test won't give any advantage right?

Comment (krajasekhar-unify) [reply src/niraj/narcissisticNumbers.spec.ts]: https://jestjs.io/docs/en/setup-teardown#scoping

Comment (nirajunify) [reply src/niraj/narcissisticNumbers.spec.ts]: Yes. What you said is absolutely correct. In this instance, I chose to use describe and test to break a long test statement.

Comment (nirajunify) [reply src/niraj/narcissisticNumbers.spec.ts]: I could have written the tests without 'describe', but adding 'describe' helped my test suits look good

Comment (krajasekhar-unify) [reply src/niraj/narcissisticNumbers.spec.ts]: Since all the tests are related to testing a single functionality, can write a single describe block to enclose all the tests.

Comment (nirajunify) [reply src/niraj/narcissisticNumbers.spec.ts]: > Since all the tests are related to testing a single functionality, can write a single describe block to enclose all the tests.

Perhaps, Can you please give recommendation on how to arrange my tests in logical grouping?

Comment (krajasekhar-unify) [reply src/niraj/narcissisticNumbers.spec.ts]: how about something like this?
```
describe('Narcissistic number', () => {
   test('0 is not a narcissistic number', () => {... });
   test('it should return an empty array for a negative starting number input', () => {... });
   ...
});
```

Comment (ramjayunify) [reply src/niraj/narcissisticNumbers.spec.ts]: describe can be used to group test cases of similar kinds. In this case @nirajunify has created a detailed test cases, each one representing different types or edge cases. May be we can group those things into a nested describe such as negative numbers in one group
```
describe('Given a negative number', () => {
  describe('Given a negative starting number', () => {
    test('it should return an empty array', () => {
      expect(narcissisticNumbers(-10).length).toBe(0);
    });
  });

  describe('Given a negative ending number', () => {
    test('it should return an empty array', () => {
      expect(narcissisticNumbers(0, -10).length).toBe(0);
    });
  });
});
```

Or when we are testing for different range of positive numbers like this

```
describe('When starting number is 10 and ending number is 10000', () => {
  const narcissisticNumbersList = [153, 370, 371, 407, 1634, 8208, 9474];
  test('it should return an array of numbers [153, 370, 371, 407, 1634, 8208, 9474]', () => {
    expect(narcissisticNumbers(10, 10000).length).toEqual(
      narcissisticNumbersList.length
    );
    expect(narcissisticNumbers(10, 10000)).toEqual(narcissisticNumbersList);
  });

  test('it should return an array of numbers [153, 370, 371, 407]', () => {
    expect(narcissisticNumbers(10, 500).length).toEqual(
      narcissisticNumbersList.splice(0,4).length
    );
    expect(narcissisticNumbers(10, 500)).toEqual(narcissisticNumbersList.splice(0,4));
  });
});
```
But I read a blog, where it says grouping test cases using describe is not the best way to do it.
Please refer this [https://kentcdodds.com/blog/avoid-nesting-when-youre-testing](url)

Comment (nirajunify) [reply src/niraj/narcissisticNumbers.spec.ts]: Thank you @krajasekhar-unify , @ramjayunify  for your valuable feedback. I will update the tests

File: src/niraj/narcissisticNumbers.ts [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+/**
+ * Function to print all the narcissistic numbers between start and end params
+ * @param {number} start - Positive number to start the counting
+ * @param {number} end - Positive number to end the counting
+ * @return {number[]} returns an array of numbers
+ */
+
+export const narcissisticNumbers = (start?: number, end?: number): number[] => {
+  const result = [];
```
Comment (praveen-unify) [inline src/niraj/narcissisticNumbers.ts]: we can define the type for 'result' as well.

File: src/niraj/narcissisticNumbers.ts [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+  if (!start || !end || start < 0 || end < 0 || end < start) return result;
+
+  for (let i = start; i <= end; i++) {
```
Comment (praveen-unify) [inline src/niraj/narcissisticNumbers.ts:42]: for 'i' also we can define it's type.

File: src/niraj/narcissisticNumbers.ts [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+ * @return {boolean}
+ */
+
+export const isNarcissisticNumber = (numberToCheck: number): boolean => {
+  if (numberToCheck <= 0) return false;
+
+  // All numbers between 1-10 are narcissistic number
+  if (numberToCheck > 0 && numberToCheck < 10) return true;
+
+  const strNum = String(numberToCheck);
```
Comment (praveen-unify) [inline src/niraj/narcissisticNumbers.ts:13]: We can define the type of 'strNum'.

File: src/niraj/narcissisticNumbers.ts [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+  return (
+    numberToCheck ==
```
Comment (praveen-unify) [inline src/niraj/narcissisticNumbers.ts]: Please use 'strict equality' using === instead of ==

**Review by praveen-unify — COMMENTED** [review]
I have some feedback to changes.

File: src/niraj/narcissisticNumbers.spec.ts (hunk 1 above)
Comment (zulqer-nain) [reply src/niraj/narcissisticNumbers.spec.ts]: I think in each describe group there must be at least one pass case and one fail case .

Comment (krajasekhar-unify) [reply src/niraj/narcissisticNumbers.spec.ts]: Thanks @ramjayunify for the `kentcdodds` article. I am convinced to not group the tests with describe at all. Do we have anything to inspire from the article and define in our guidelines?

File: src/niraj/narcissisticNumbers.ts (hunk 4 above)
Comment (krajasekhar-unify) [reply src/niraj/narcissisticNumbers.ts:13]: Please help to clarify my doubt. Do we need to do type check for every intermediate value?
- If we are using any js functions which always outputs a particular type (String -> string, Math.sqrt -> number), it already knows the type right?
- For the below code, even though type is not declared for `varB` and `varA`, at the time of using it editor is still showing the type as number as it reflected from the return type of the function? Do we still need to define types for immediate variables?
```
const myFunction = (): number => 4;
const varB = myFunction();
const varA = varB + 3;
console.log(varA, varB);
```

Comment (nirajunify) [reply src/niraj/narcissisticNumbers.ts:13]: You are right Raj. Wherever typescript can automatically infer a variable's type, you do not have to provide its type.

File: src/niraj/narcissisticNumbers.ts (hunk 3 above)
Comment (nirajunify) [reply src/niraj/narcissisticNumbers.ts:42]: No necessarily. Since I am defining the variable and immediately assigning the value, typescript can automatically infer that variable's type - I don't need to define the type.

File: src/niraj/narcissisticNumbers.ts (hunk 2 above)
Comment (nirajunify) [reply src/niraj/narcissisticNumbers.ts]: Agreed. I will update the code

File: src/niraj/narcissisticNumbers.spec.ts (hunk 1 above)
Comment (nirajunify) [reply src/niraj/narcissisticNumbers.spec.ts]: Thanks. I will update the specs

File: .gitignore [hunk 6]
```diff
@@ -103,3 +103,8 @@ dist
 # TernJS port file
 .tern-port
 
+<<<<<<< HEAD
+src/**/*.map
+coverage
+=======
+>>>>>>> 5d7f7554696eab1be0365ee5a4ca666322d36243
```
Comment (praveen-unify) [inline .gitignore]: Conflict is not resolved, please resolve it.

Comment (nirajunify) [reply .gitignore]: It's been resolved now

**Review by praveen-unify — DISMISSED** [review]

**Review by krajasekhar-unify — APPROVED** [review]
